onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -430,14 +464,17 @@ Status CUDAExecutionProvider::OnRunEnd(bool sync_stream, const onnxruntime::RunO
 }
 
 bool CUDAExecutionProvider::IsGraphCaptureEnabled() const {
-  return info_.enable_cuda_graph;
+  return info_.enable_cuda_graph && GetPerThreadContext().IsGraphCaptureAllowedOnRun();
 }
 
-bool CUDAExecutionProvider::IsGraphCaptured() const {
+bool CUDAExecutionProvider::IsGraphCaptured(int graph_annotation_id) const {
+  ORT_ENFORCE(GetPerThreadContext().SetCudaGraphAnnotationId(graph_annotation_id) == Status::OK());
```

Comment:
I think the id is set during OnRunStart.
May change to assert that the cuda graph id in context is name as graph_annotation_id.

Reply:
If graph is captured in previous run, the two functions(IsGraphCaptured() and ReplayGraph()) are called before OnRunStart: https://github.com/microsoft/onnxruntime/blob/0281e570f9e3dd709a94d4d5f99604f905abc07e/onnxruntime/core/session/inference_session.cc#L2403